[PL-Doc] OIDC SSO with Okta & Harness (#9442)
harness/developer-hub: docs/platform/authentication/single-sign-on-sso-with-oidc.md

Reviewer: Bullets used are not consistent. Either use a,b,c for sub-bullets or use bullet icon.

Author: For header i have used 1,2,3 and bullets for sub header

---
title: Single Sign-On (SSO) with OpenID Connect (OIDC) 
description: This document explains single sign-on with OIDC provider.
# sidebar_position: 2
helpdocs_topic_id: mlpksc7s6c
helpdocs_category_id: fe0577j8ie
helpdocs_is_private: false
helpdocs_is_published: true
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

:::warning Important Note
    - The OIDC authentication feature is controlled by the feature flag `PL_ENABLE_OIDC_AUTHENTICATION`, which must be enabled to use this feature.
    - OIDC Authentication is only supported for accounts with **[Vanity URL](https://developer.harness.io/docs/platform/authentication/authentication-overview#set-up-vanity-url)**
:::

> Harness supports Single Sign-On (SSO) with any custom OIDC (OpenID Connect) provider, allowing authentication and provisioning user and user group. Currently, only the Authorization Code flow is supported, enabling applications to exchange an authorization code for a token.

Reviewer: Authorization code flow pointer should be highlighted as a Note

### Getting Started

Before you begin, It's helpful to check the following:

1. Review the **[Authentication Overview](https://developer.harness.io/docs/platform/authentication/authentication-overview)** in Harness.  
2. Check how **[Role-Based Access Control (RBAC)](https://developer.harness.io/docs/platform/role-based-access-control/rbac-in-harness)** works.  
3. Learn more about **[OpenID Connect (OIDC)](https://openid.net/developers/how-connect-works/)** by referring to the official documentation.

### Supported Features

- **OIDC-Based Authentication** – Authenticate users using the OIDC protocol.  
- **Multiple OIDC Providers** – Configure and manage multiple OIDC identity providers.  
- **Just-in-Time (JIT) User Provisioning** – Automatically create and update users upon login via OIDC.  
- **User Group Provisioning** – Assign users to groups based on their OIDC provider group membership.  

### OIDC SSO with Okta 

To configure Harness with Okta for OIDC SSO, you need to exchange required information between your **Okta application** and **Harness**.

> This section outlines the steps to integrate Okta as an OIDC provider for seamless authentication in Harness. 

Reviewer: Add the Tip mentioned in Google doc.

Author: Added.

### Steps to Create an App Integration in Okta

1. **Sign in to Okta**  
   - Log in to your Okta **administrator account**.  
   - Navigate to **Applications** and click **Create App Integration**.     
     
     ![create-app-okta](./static/create-app-integration.png)

> 2. **Select Sign-in Method**  

Reviewer: This step covers selecting both Sign-in Method and application type. The title **Select Sign-in Method**   does not capture that.

Author: Added **Application Type**

   - In the **Create a new app integration** dialog, choose **OIDC - OpenID Connect** as the sign-in method.  

>    - Select **Web Application** as the application type.

Reviewer: ```suggestion
   - Select **Web Application** as Application type.
```

Reviewer: Keeping it consistent with Okta

     
     ![sign-in-method-okta](./static/sign-in-method-okta.png)  

3. **Configure General Settings**  
   - Enter a name for the app in **App Integration Name**.  
   - Under **Sign-in redirect URIs**, enter your **Harness URL** followed by:  
     ```

>      /gateway/user/auth/oidc/callback

Reviewer: let us keep this as inline code rather than a code block as otherwise it will consume unnecessary space.

Author: Make sense. updated.

     ```
     **Example:** `https://something.harness.io/gateway/user/auth/oidc/callback`

4. **Set User Access**  
   - Under **Assignments**, select the **user groups** that should have access to the application.  
   - Other details are optional.  

5. **Save and Proceed**  
   - Click **Save** to complete the app integration setup.
        

>     ![step-3-4-5-config-okta](./static/newweb-oidc-app.gif)

Reviewer: Comments for Gif:
1. Sign in-redirect URI should have value: https://something.harness.io/gateway/user/auth/oidc/callback as mentioned in the example above
2. Remove the sign out URL from gif.

Author: updated.

### Adding an Okta OIDC Provider in Harness

#### Step 1: Configure the OIDC Provider

1. Navigate to the **Account Settings ➝ Security and Governance ➝ Authentication** page in Harness.
    

>     Check and select  **OIDC Provider** 

Reviewer: This should be: Select Login via OIDC and click on Add OIDC Provider. This should be consistent with the Product labels and experience. Also, this point should be a sub-bullet.

>     ![authentication-oidc](./static/authentication-oidc.gif)

Reviewer: Update the Gif as per the steps mentioned above. Also, it would be great if we can show a fresh experience here to the user where no OIDC provider is existing. The current GIF already has 4 providers.

Author: This is bit unfeasible let me check with devs if they can provide.

2. Configure OIDC 
    - Enter a **name** for the OIDC configuration.  

>     - In **OIDC Scope**, the default required values (`openid, email, profile`) are pre-selected. You may add additional scopes if needed.  

Reviewer: ```suggestion
    - In **OIDC Scope**, the default required values (`openid`, `email`, `profile`) are pre-selected. You may add additional scopes if needed.  
```

>     - Under **Issuer**, enter your **Okta URL** (e.g., `https://example-123.oktapreview.com`).  

Reviewer: ```suggestion
    - Under **Issuer**, enter the Issuer URL from your authorization server (e.g., `https://example-123.oktapreview.com`).  
```

>     - In **UID Field**, enter `email` to use the email address as the unique identifier.  

Reviewer: We mandatorily require the attribute that captures the email value. Others cannot work. We have to write this in such  a way that this idea gets captured. 

Suggestion: In **UID Field**, add the attribute that contains the email address of the user. In most of the cases, it is `email. ' Only email address is supported as the unique identifier.

Author: @abhishekharness this works?

> In the **UID Field**, enter the attribute that contains the user's email address. In most cases, this is `email`. **Only email addresses are supported as the unique identifier, and other attributes will not work.**

Reviewer: We can just mention that only email addresses are supported and skip other attributes won't work part.

    - Click **Continue** to proceed. 
    
        ![configure-oidc](./static/oidc-page-1.png)

#### Step 2: Configure Client Settings

Client Settings include the credentials and endpoints needed for Harness to connect with the OIDC provider(Okta).

<Tabs>
  <TabItem value="DE" label="Discovery Enabled (Default)" default>

>         - If **Discovery is enabled** (default behavior):  

Reviewer: This should not be a bullet point

>         - Harness automatically retrieves Identity Provider (IdP) details from the `/.well-known` endpoint. (i.e., `https://example-123.oktapreview.com/.well-known/openid-configuration` )  

Reviewer: This should be clubbed with the previous line

        - Provide the following:  

>             - **Client Identifier** – Found in Okta under "Client Credentials" in the General tab.  

Author: updated as per doc

>             - **Client Secret** – Found in the same section.  

Reviewer 2: @ivikramsahu Can you specify here that this secret should necessarily be a part of the Built-In Harness Secrets Manager?

Author: Ack & done.

>             - **Client Secret** – Found in the same section.  

Author: updated.

>             - **Redirect URL** – Must match the Sign-in Redirect URI from the Okta app setup (e.g., `https://something.harness.io/gateway/user/auth/oidc/callback`). 

Author: resolved

                ![oidc-with-discovery](./static/oidc-with-discovery.png) 
    </TabItem>
   <TabItem value="DD" label="Discovery Disabled">

>         - If **Discovery is disabled**, enter additional details:  

Reviewer: This should not be a bullet point.

        - **Authorization Endpoint** – URL for user authorization (e.g., `https://example-123.oktapreview.com/oauth2/v1/authorize`).  
        - **Token Endpoint** – URL for retrieving access tokens (e.g., `https://example-123.oktapreview.com/oauth2/v1/token`).  
        - **User Info Endpoint** – URL for obtaining user details (e.g., `https://example-123.oktapreview.com/oauth2/v1/userinfo`).  
        - **JWKS URI** – URL where the token signer publishes its keys (e.g., `https://example-123.oktapreview.com/oauth2/v1/keys`).
            
            ![oidc-without-discovery](./static/oidc-without-discovery.png)              
    </TabItem>
</Tabs>

Click on **Continue** to proceed

#### Step 3: Additional Settings (Optional)

> **1. Enable JIT Provisioning**

Reviewer: @ivikramsahu this is still not resolved.

> **1. Enable JIT Provisioning**

Author: i have noted this in Google Doc

        ![JIT-provisioning](./static/claim-key-value-jit.png)  

    Additionally, **Claim Key** and **Claim Value** help control automatic provisioning. If these values match specific attributes in the **ID token** received from the OIDC provider, the user is automatically created in Harness.

    Refer to the **[Okta documentation](https://developer.okta.com/docs/guides/customize-tokens-returned-from-okta/main/)** for guidance on customizing Okta tokens with custom claims.

**2. Authorization is enabled**

    You can enable Okta OIDC authorization in Harness by linking a **[Harness user group](https://developer.harness.io/docs/platform/role-based-access-control/add-user-groups)** to an **Okta user group**. When a user from the linked Okta group logs into Harness, they are automatically added to the corresponding Harness user group and inherit its permissions and access.  

    For more details, refer to the **[RBAC section](https://developer.harness.io/docs/platform/role-based-access-control/rbac-in-harness)** in Harness documentation.

    1. Setting Up OIDC SSO and Authorization in Harness

>         1. Enabling OIDC authorization allows the same users authenticated via your OIDC provider to be authorized in Harness.  

Author: removed from step kept it as statement info

            - In the **Add Group** dialog, enter a **Group Name** and **Description**, then click **Add Group**.

                ![user-group in okta](./static/okta-user-group.png)

        3. **Add Users to the Group**:  
            - After creating the group, search for it in the **Groups** section.  
            - Select the group and click **Assign People**.  
            - Search for users and add them to the group.  

                ![Okta-user-group](./static/okta-group-assignment.gif)

        4. **Verify Group Members**:  
            - After adding members, the group should display the list of users added.

:::tip
    Make sure to **note the Okta Group Name**, as you’ll need it later to link the Okta group to a Harness user group.
:::

    2. Ensure the **Okta user group** is assigned to the same Okta OIDC provider app you use for **Harness OIDC SSO**. Here’s how to do it:

        1. In Okta, go to **Directory** > **Groups** and select your Okta user group.
        2. Click **Assign Apps** and find the your **Okta app**.
        3. Select **Assign** and click **Done**.
            
            ![app-assignment](./static/okta-assign-application.gif)

>         4. Then, under **Applications**, go to **Applications** and select the **Okta app** you created in previous step.

Author: make sense

        5. In the **Assignments** tab, check that your Okta user group is listed.
            
            ![okta-app-assignments](./static/okta-app-assignments.png)

>     3. Create a **Groups Claim** for your OpenID Connect client app by following the steps in the official Okta documentation. This claim will later be used to enable **OIDC authorization** in Harness.

Reviewer: Link to Okta documentation is missing. Please refer the Google doc.

        1. In Okta, go to **Applications** and select your **Harness Okta OIDC SSO app**.
        2. Go to the **Sign On** tab and click **Edit** under the **OpenID Connect ID Token** section.

>         3. Use the **Groups claim filter** or **Groups claim expression** to list the user’s groups. The filter allows you to select which groups should be authenticated to Harness.  

Reviewer: ```suggestion
        3. Use the **Groups claim filter** or **Groups claim expression** that containes the list of user’s groups. The filter allows you to select which groups should be authenticated to Harness.  
```

>             - In the **Groups claim filter** section, make sure the name "groups" is present (add it if empty).  

Reviewer: ```suggestion
            - In the **Groups claim filter** section, make sure the name "groups" is present (add if it is empty).  
```

>             - Set the filter type to **Matches regex** and enter `.*` to return all user groups.

Reviewer: This is not a step and only applicable in case regex is used. Please combine with the previous step.

        4. Click **Save** to complete the configuration.

            ![group-assign-okta-application](./static/group-assign-okta-application.gif)

    4. Enable Group Claim Authorization in Harness

        1. In **Harness**, go to **Account Settings** > **Authentication**.  
        2. Expand the **Login via OIDC** section.  
        3. Click **More options (⋮)** next to your **Okta provider configuration** and select **Edit**.  
            ![check-edit-config](./static/edit-for-okta-provider.png)

>         4. On the **OIDC Provider Overview** page, add **groups** as an additional **OIDC Scope** if using an **Org authorization server** in Issuer.  

Reviewer 2: @ivikramsahu can we also add a screenshot with the added scope- "groups" so that it is not missed during the setup?

Author: done added

        5. In **Additional Settings**, enable **Authorization**.  
        6. Set **Group Claim** as `groups`.  
        7. Click **Submit** to save changes. Now, your **Okta configuration** uses the **Group Claim** for authorization.
            ![Auth-enabled-group](./static/auth-enabled.png)
    

> ### Link Okta User Group to a Harness User Group

Reviewer: This is also under "Authorization is enabled"

    1. In **Harness**, go to **Account Settings** > **Access Control**.  
    2. Click on **User Groups** and find the group you want to link with your Okta user group.  
    3. Select **Link to SSO Provider Group**.  

>     4. In the **Search SSO Settings** window, choose your **Okta OIDC SSO configuration**.  

Reviewer: ```suggestion
    4. In the **Search SSO Settings** window, choose your Okta OIDC SSO configuration.  
```

Reviewer: We generally used Bold only for fields that are available in the Product.

    5. Enter the **Okta Group Name** and click **Save**.

>         ![link-to-sso](./static/link-to-sso.gif)

Reviewer: Please add customer-friendly names

Author: will fix this in next draft.

    6. Repeat these steps for any additional user groups you need to connect.

> ### Test and Verify OIDC Authorization in Harness

Reviewer: This should be a sub-section in the previous section as this is not independent.

    1. **Log in via OIDC:**  
        - Open a **private/incognito browsing window** and go to **Harness**.  
        - Log in using a **Harness user account** with an email registered in **Okta**.  
        - If configured correctly, you'll be redirected to the **Okta login page**. 

    2. **Authenticate with Okta:**  
        - Enter your **email address** (passwords for Harness and Okta can be different).  
        - If authentication is successful, you’ll be redirected back to **Harness**.
            

>             > **Note:** The login screen below will appear only if you have multiple OIDC configurations set up.

Reviewer: This note is incorrect.

Author: removed note part from this.

### Steps to Create an App Integration in Okta

1. **Sign in to Okta**  
   - Log in to your Okta **administrator account**.  
   - Navigate to **Applications** and click **Create App Integration**.     
     
     ![create-app-okta](./static/create-app-integration.png)

2. **Select Sign-in Method and Application Type**  
   - In the **Create a new app integration** dialog, choose **OIDC - OpenID Connect** as the sign-in method.  
   - Select **Web Application** as application type.
     
     ![sign-in-method-okta](./static/sign-in-method-okta.png)  

3. **Configure General Settings**  
   - Enter a name for the app in **App Integration Name**.  
   - Under **Sign-in redirect URIs**, enter your **Harness URL** followed by `/gateway/user/auth/oidc/callback` 
   
    **Example:** `https://something.harness.io/gateway/user/auth/oidc/callback`

4. **Set User Access**  
   - Under **Assignments**, select the **user groups** that should have access to the application.  
   - Other details are optional.  

5. **Save and Proceed**  
   - Click **Save** to complete the app integration setup.
        

>     ![step-3-4-5-config-okta](./static/newweb-oidc-app.gif)

Reviewer: @ivikramsahu The GIF at the end shows a a blank screen. Also, there is no capability to replay these GIFs. Doesn't seem to be a good user experience. Is it possible to provide ability to replay them?

Author: let's replace then in upcoming draft

### Adding an Okta OIDC Provider in Harness

#### Step 1: Configure the OIDC Provider

1. Navigate to the **Account Settings ➝ Security and Governance ➝ Authentication** page in Harness.
    
    - Select Login via OIDC and click on Add OIDC Provider 

>         ![authentication-oidc](./static/authentication-oidc.gif)

Reviewer: @ivikramsahu as commented earlier, the GIF should show a fresh start without any OIDC Provider

Author: updated gif

2. Configure OIDC 
    - Enter a **name** for the OIDC configuration.  
    - In **OIDC Scope**, the default required values (`openid`, `email`, `profile`) are pre-selected. You may add additional scopes if needed.  
    - Under **Issuer**, enter the Issuer URL from your authorization server (e.g., `https://example-123.oktapreview.com`).  
    - In **UID Field**, enter `email` to use the email address as the unique identifier.  
    - Click **Continue** to proceed. 
    
        ![configure-oidc](./static/oidc-page-1.png)

#### Step 2: Configure Client Settings

Client Settings include the credentials and endpoints needed for Harness to connect with the OIDC provider(Okta).

<Tabs>
  <TabItem value="DE" label="Discovery Enabled (Default)" default>
        If **Discovery is enabled** (default behavior), Harness automatically retrieves Identity Provider (IdP) details from the `/.well-known` endpoint. (i.e., `https://example-123.oktapreview.com/.well-known/openid-configuration` )  

>         - Provide the following:  

Reviewer: This line is not really required. We can combine this with the above statement itself to avoid bullets and then sub-bullet. We have used the same approach in Discovery = false. Let's keep both of them consistent.

    Refer to the **[Okta documentation](https://developer.okta.com/docs/guides/customize-tokens-returned-from-okta/main/)** for guidance on customizing Okta tokens with custom claims.

**2. Authorization is enabled**

    You can enable Okta OIDC authorization in Harness by linking a **[Harness user group](https://developer.harness.io/docs/platform/role-based-access-control/add-user-groups)** to an **Okta user group**. When a user from the linked Okta group logs into Harness, they are automatically added to the corresponding Harness user group and inherit its permissions and access.  

    For more details, refer to the **[RBAC section](https://developer.harness.io/docs/platform/role-based-access-control/rbac-in-harness)** in Harness documentation.

    1. Setting up Authorization for an OIDC provider in Harness

        Enabling OIDC authorization allows the same users authenticated via your OIDC provider to be authorized in Harness.  

        1. **Create a User Group in Okta** (if not already created):  
            - Log in to **Okta Admin Account**.  

>             - Go to **Directory** > **Groups**, then click **Add Group**.  

Reviewer: ```suggestion
            - Go to **Directory** > **Groups**, then click **Add group**.  
```

                ![user-group in okta](./static/okta-user-group.png)

        2. **Add Users to the Group**:  
            - After creating the group, search for it in the **Groups** section.  
            - Select the group and click **Assign people**.  
            - Search for users and add them to the group.  

                ![Okta-user-group](./static/okta-group-assignment.gif)

        3. **Verify Group Members**:  
            - After adding members, the group should display the list of users added.

:::tip
    Make sure to **note the Okta Group Name**, as you’ll need it later to link the Okta group to a Harness user group.
:::

    2. Ensure the **Okta user group** is assigned to the same Okta OIDC provider app you use for **Harness OIDC SSO**. Here’s how to do it:

        1. In Okta, go to **Directory** > **Groups** and select your Okta user group.
        2. Go to the **Applications** tab and click on **Assign applications**. Find the Okta application created by you.
        3. Select **Assign** and click **Done**.
            
            ![app-assignment](./static/okta-assign-application.gif)

>         4. Go to the **Applications** page and select the Okta app you created in previously.

Reviewer: ```suggestion
        4. Go to the **Applications** page and select the Okta app you created previously.
```

        1. In Okta, go to **Applications** and select your **Harness Okta OIDC SSO app**.
        2. Go to the **Sign On** tab and click **Edit** under the **OpenID Connect ID Token** section.
        3. Use the **Groups claim filter** or **Groups claim expression** that contains the list of user’s groups. The filter allows you to select which groups should be authenticated to Harness. In the **Groups claim filter** section, make sure the name "groups" is present (add if it is empty). Set the filter type to **Matches regex** and enter `.*` to return all user groups.
        4. Click **Save** to complete the configuration.

            ![group-assign-okta-application](./static/group-assign-okta-application.gif)

    4. Enable Group Claim Authorization in Harness

        1. In **Harness**, go to **Account Settings** > **Authentication**.  
        2. Expand the **Login via OIDC** section.  
        3. Click **More options (⋮)** next to your Okta provider configuration and select **Edit**.  
            ![check-edit-config](./static/edit-for-okta-provider.png)

>         4. On the **OIDC Provider Overview** page, add **groups** as an additional **OIDC Scope** if using an **Org authorization server** in Issuer.  

Reviewer: ```suggestion
        4. On the **OIDC Provider Overview** page, add **groups** as the additional **OIDC Scope** if using an **Org authorization server** in Issuer.  
```

Reviewer: There is just one Org authorization server. Hence, suggested the above change.

            ![oidc-scope](./static/oidc-groups-scope.png)
        5. In **Additional Settings**, enable **Authorization**.  
        6. Set **Group Claim** as `groups`.  
        7. Click **Submit** to save changes. Now, your **Okta configuration** uses the **Group Claim** for authorization.
            ![Auth-enabled-group](./static/auth-enabled.png)
    
    5. Link Okta User Group to a Harness User Group

        1. In **Harness**, go to **Account Settings** > **Access Control**.  
        2. Click on **User Groups** and find the group you want to link with your Okta user group.  
        3. Select **Link to SSO Provider Group**.  
        4. In the **Search SSO Settings** window, choose your **Okta OIDC SSO configuration**.  
        5. Enter the **Okta Group Name** and click **Save**.

>             ![link-to-sso](./static/link-to-sso.gif)

Reviewer: Same comment as before. This should be a customer friendly name.

Author: added in google doc
